data-serving/scripts/converters.py

```diff
@@ -0,0 +1,121 @@
+'''Library of conversion functions for converting CSV line-list data to json.'''
+
+import datetime
+import logging
+import math
+import numbers
+import pandas as pd
+from typing import Any, Callable, Dict, List
+
+
+VALID_SEXES = ['female', 'male', 'other']
+
+range_ct = 0
+single_ct = 1
+
+
+def convert_range(value: str, parse_fn: Callable[[str], str], format_fn: Callable[[Any], Any]):
+    '''
+    Converts either a string representation of a single value or a range of values into a range object. The values can be of any types using the custom parse and format functions.
+    '''
+    if pd.isna(value):
+        return None
+
+    # First check if the value is represented as a range.
+    if type(value) is str and value.find('-') >= 0:
+        value_range = value.split('-')
+        # Remove whitespace that might surround the dash ('23 - 72')
+        value_range = [x.strip() for x in value_range]
+        # Remove empty strings that can result from open ranges ('18-')
+        value_range = [i for i in value_range if i]
+
+        # Handle open ranges (i.e. missing min or max).
+        range_min = parse_fn(value_range[0])
+        if len(value_range) == 1:
+            return {
+                'range': {
+                    'min': format_fn(range_min)
+                }
+            } if value.startswith('-') else {
+                'range': {
+                    'max': format_fn(range_min)
+                }
+            }
+
+        # Handle cases with a min and max.
+        range_max = parse_fn(value_range[1])
+        return {
+            'range': {
+                'min': format_fn(range_min),
+                'max': format_fn(range_max)
+            }
+        }
+
+    # We have a specific value. We create a range with identical min and max values.
+    parsed_value = parse_fn(value)
+    return {
+        'range': {
+            'min': format_fn(parsed_value),
+            'max': format_fn(parsed_value)
+        }
+    }
+
+
+def convert_event(id: str, name: str, date_str: str) -> Dict[str, str]:
+    '''
+    Converts a single event date column to the new event object with a name and range of dates.
+    '''
+    if pd.isna(date_str):
+        return None
+
+    try:
+        date_range = convert_range(date_str, lambda x: datetime.datetime.strptime(
+            x, '%d.%m.%Y'), lambda x: {
+                '$date': x.strftime('%Y-%m-%dT%H:%M:%SZ')
+        })
+
+        return {
+            'name': name,
+            'date': date_range
+        }
+    except (TypeError, ValueError):
+        logging.warning(
+            '[%s] [event[name=.%s]] invalid value %s', id, name, date_str)
+
+
+def convert_events(id: str, dates: str, outcome: str) -> List[Dict[str, Any]]:
```

Comment:
I thought these were tuples? E.g. we'd be returning `[ {name: 'eventName1', date: $date1}, ...]`? I must be missing a layer.

Reply:
not sure I follow -- wouldn't that example you have there be a list of dictionaries (objects)?

Reply:
I think I'm mixing metaphors (and by metaphors I mean punctuation). I meant to ask: is it better to represent that data as a dict, a tuple of [str, date], or a defined class/object? In the grand scheme of things, if the dict approach is working for you thus far, it's probably good enough.

Reply:
oh, gotcha. I think I chose dict because it's the most json-y, so it's easier for me to scan and visualize the output -- but I'm open to other approaches!